src/common/token_bucket.c

```diff
@@ -0,0 +1,172 @@
+/* Copyright (c) 2003-2004, Roger Dingledine
+ * Copyright (c) 2004-2006, Roger Dingledine, Nick Mathewson.
+ * Copyright (c) 2007-2017, The Tor Project, Inc. */
+/* See LICENSE for licensing information */
+
+/**
+ * \file address.c
+ * \brief Functions to use and manipulate token buckets, used for
+ *    rate-limiting on connections and globally.
+ *
+ * Tor uses these token buckets to keep track of bandwidth usage, and
+ * sometimes other things too.
+ *
+ * The time units we use internally are based on "timestamp" units -- see
+ * monotime_coarse_to_stamp() for a rationale.
+ *
+ * Token buckets may become negative.
+ **/
+
+#define TOKEN_BUCKET_PRIVATE
+
+#include "token_bucket.h"
+#include "util_bug.h"
+
+/** Convert a rate in bytes per second to a rate in bytes per step */
+static uint32_t
+rate_per_sec_to_rate_per_step(uint32_t rate)
+{
+  return rate / (STAMP_TICKS_PER_SECOND / TICKS_PER_STEP);
+}
+
+/**
+ * Initialize a token bucket in *<b>bucket</b>, set up to allow <b>rate</b>
+ * bytes per second, with a maximum burst of <b>burst</b> bytes. The bucket
+ * is created such that <b>now_ts</b> is the current timestamp.  The bucket
+ * starts out full.
+ */
+void
+token_bucket_init(token_bucket_t *bucket,
+                  uint32_t rate,
+                  uint32_t burst,
+                  uint32_t now_ts)
```

Comment:
Within the DoS subsystem, I did something extremely similar (of course bucketting...) but I went for `uint64_t` on purposes so we could address just all use cases in the future. Turned out that we could just have `uint32_t` instead since we use consensus param for some values which are bound to be `int32_t`.

You think we could go to 64 bit there or memory footprint is the idea of 32 bit?

Reply:
Hm.  So, the existing code here assumes it's 32-bit.

Do we ever need to allow a burst of greater than 4GB?  Nobody has actually asked for it so far.  And since the API returns a size_t, it should be comparatively easy to make this bigger later on if we want it.

Reply:
Ok, after seeing how it is used in the code and tightly coupled with torrc options that are int32_t, this is fine.